cardano-cli/src/Cardano/CLI/Run/Friendly.hs

```diff
@@ -11,167 +11,203 @@
+                   getTransactionBodyContent, multiAssetSupportedInEra, serialiseAddress,
+                   serialiseAddressForTxOut, txMetadataSupportedInEra, updateProposalSupportedInEra,
+                   validityLowerBoundSupportedInEra, validityUpperBoundSupportedInEra,
+                   withdrawalsSupportedInEra)
+import           Cardano.Api.Byron (Lovelace (..))
+import           Cardano.Api.Shelley (Address (ShelleyAddress), StakeAddress (..))
 import qualified Shelley.Spec.Ledger.API as Shelley
 
 import           Cardano.CLI.Helpers (textShow)
 
-friendlyTxBodyBS :: CardanoEra era -> Api.TxBody era -> ByteString
-friendlyTxBodyBS era =
-  encodePretty (setConfCompare compare defConfig) . friendlyTxBody era
-
-friendlyTxBody :: CardanoEra era -> Api.TxBody era -> Value
-friendlyTxBody era txbody =
-  Object $
-    HashMap.fromList ["era" .= toJSON era]
-    <>
-    case txbody of
-      ByronTxBody body -> friendlyTxBodyByron body
-      ShelleyTxBody ShelleyBasedEraShelley body _scripts aux ->
-        addAuxData aux $ friendlyTxBodyShelley body
-      ShelleyTxBody ShelleyBasedEraAllegra body _scripts aux ->
-        addAuxData aux $ friendlyTxBodyAllegra body
-      ShelleyTxBody ShelleyBasedEraMary body _scripts aux ->
-        addAuxData aux $ friendlyTxBodyMary body
-
-addAuxData :: Show a => Maybe a -> Object -> Object
-addAuxData = HashMap.insert "auxiliary data" . maybe Null (toJSON . textShow)
-
-friendlyTxBodyByron :: Annotated Byron.Tx ByteString -> Object
-friendlyTxBodyByron = assertObject . toJSON
-
-friendlyTxBodyShelley
-  :: Shelley.TxBody (Ledger.ShelleyEra StandardCrypto) -> Object
-friendlyTxBodyShelley body =
-  HashMap.fromList
-    [ "inputs" .= Shelley._inputs body
-    , "outputs" .= fmap friendlyTxOutShelley (Shelley._outputs body)
-    , "certificates" .= fmap textShow (Shelley._certs body)
-    , "withdrawals" .= Shelley.unWdrl (Shelley._wdrls body)
-    , "fee" .= Shelley._txfee body
-    , "time to live" .= Shelley._ttl body
-    , "update" .= fmap textShow (Shelley._txUpdate body)
-    , "metadata hash" .= fmap textShow (Shelley._mdHash body)
-    ]
-
-friendlyTxBodyAllegra
-  :: ShelleyMA.TxBody (ShelleyMAEra 'Allegra StandardCrypto) -> Object
-friendlyTxBodyAllegra
-  (ShelleyMA.TxBody
-    inputs
-    outputs
-    certificates
-    (Shelley.Wdrl withdrawals)
-    txfee
-    validity
-    update
-    adHash
-    _mint -- mint is not used in Allegra, only in Mary+
-    ) =
-  HashMap.fromList
-    [ "inputs" .= inputs
-    , "outputs" .= fmap friendlyTxOutAllegra outputs
-    , "certificates" .= fmap textShow certificates
-    , "withdrawals" .= withdrawals
-    , "fee" .= txfee
-    , "validity interval" .= friendlyValidityInterval validity
-    , "update" .= fmap textShow update
-    , "auxiliary data hash" .= fmap textShow adHash
+friendlyTxBodyBS
+  :: IsCardanoEra era => Api.TxBody era -> Either Prelude.String ByteString
```

Comment:
Why the change from the original function that requires `CardanoEra era`?  We prefer value level evidence for the era. Likewise for the rest of the functions in this module.

Reply:
sorry, previously changed. fixed like everywhere